engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java

```diff
@@ -0,0 +1,113 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.interop.*;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
+
+@ExportLibrary(InteropLibrary.class)
+public class ArraySlice implements TruffleObject {
+  private final Object storage;
+  private final long start;
+  private final long length;
+
+  public ArraySlice(Object storage, long start, long length) {
+    if (storage instanceof ArraySlice slice) {
+      this.storage = slice.storage;
+      this.start = slice.start + start;
+      this.length = length;
+    } else {
+      if (CompilerDirectives.inInterpreter()) {
+        if (!InteropLibrary.getUncached().hasArrayElements(storage)) {
+          throw new IllegalStateException("Vector needs array-like delegate, but got: " + storage);
+        }
+      }
```

Comment:
Anyway, if left as-is I think it would be beneficial to add a short comment explaining why this is only checked in interpreter.

Unless there is some obvious reason that every Truffle developer should understand immediately, I think such an explanation will help future developers understand this decision.

Reply:
We could do the check as `assert`. Asserts are ignored when Truffle does its PE compilation. We just need to make sure we run our tests with `-ea`.

Reply:
Why can't the check be there _always_?